refactor(schematics):entity schematics should update the state to plural

santoshyadav198613 opened the pull request:
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngrx/platform/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Documentation has been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
currently the state variable is singular updating the same to plural
Closes #1412

## What is the new behavior?

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

## Other information

santoshyadav198613 commented:
@brandonroberts  ,

Need help here don't know why CI is getting failed for e2e testing.

timdeschryver commented:
@santoshyadav198613 , the failing e2e test isn't related to your change.
I will take a look at it.

timdeschryver commented:
The problem is that we are caching the cypress installation in the wrong build step.
This should be solved when #1593 gets merged.

santoshyadav198613 commented:
> d it needs some tests for common scenarios.

But this code is part of common function inside schematics-core , can you guide me how should we handle it?

brandonroberts commented:
You create a test for a generated schematic and check that the output is what you expect. For example, your test would use the entity schematic to generate a `user` collection. The schematic should generate a property named `users`, with the associated files.

santoshyadav198613 commented:
> Should we use the `pluralize` method at this point?
> Because I'm a bit afraid this would lead to inconsistencies, for example, while scaffolding the effects we append the name with an **s**.

Ok let me check it.

santoshyadav198613 commented:
> You create a test for a generated schematic and check that the output is what you expect. For example, your test would use the entity schematic to generate a `user` collection. The schematic should generate a property named `users`, with the associated files.

ok will update and let you know.

santoshyadav198613 commented:
> You create a test for a generated schematic and check that the output is what you expect. For example, your test would use the entity schematic to generate a `user` collection. The schematic should generate a property named `users`, with the associated files.

Hi @brandonroberts ,

Added a test case for the scenario this is what i have done.
Added one store and entity and then verified if state variable is plural or not same for reducers.

santoshyadav198613 commented:
@brandonroberts and @timdeschryver let me know about the usage of plural function should i remove it.

santoshyadav198613 commented:
Hi @brandonroberts ,
Can you verify again, wrote the test case as per our discussion.

brandonroberts commented:
@santoshyadav198613 we don't want to modify existing behavior for other areas. Instead, add an `options` object to the ngrx-utils methods as the second argument that's set to { pluralize: false }`. For entity schematics, pass `{ pluralize: true }` and use it whether to call `pluralize` or just `camelize` inside those functions.

timdeschryver commented:
I'm a bit hesitant to add `pluralize` (for now) because this would lead to inconsistencies, e.g. city would be pluralized to cities, and in the other schematics to citys. 

What about keeping this PR scoped to making the entities pluralized (adding an `s`), and to create a separate issue to pluralize all the schematics via this function?

brandonroberts commented:
@timdeschryver yes, scoping this to entities is what I had in mind also.

santoshyadav198613 commented:
> @santoshyadav198613 we don't want to modify existing behavior for other areas. Instead, add an `options` object to the ngrx-utils methods as the second argument that's set to { pluralize: false }`. For entity schematics, pass `{ pluralize: true }`and use it whether to call`pluralize`or just`camelize` inside those functions.

Thanks @brandonroberts ,
Will make the changes.

santoshyadav198613 commented:
Hi @brandonroberts and @timdeschryver ,
Done with the changes as per discussion, test cases are also added/modified.
The change is restricted to entity only, also appending only 's', keeping pluralize let me know if i should remove it.

timdeschryver commented on modules/schematics/src/entity/index.spec.ts:
```diff
@@ -159,4 +161,34 @@ describe('Entity Schematic', () => {
     ).toBeGreaterThanOrEqual(0);
   });
+
+  it('should create all files of an entity within grouped and nested folders', () => {
```
This spec name has to be updated

santoshyadav198613 replied:
done

timdeschryver commented on modules/schematics-core/utility/ngrx-utils.ts:
```diff
@@ -152,11 +154,13 @@ export function addReducerToActionReducerMap(
   let node = actionReducerMap.initializer;
 
+  const state =
+    options.plural === undefined
```
This check isn't sufficient, if we set `plural` to `false` we would still pluralize the name and this might be confusing 😄

santoshyadav198613 replied:
Yes i was thinking about same, after i went to sleep. thanks for catching it.

timdeschryver commented on modules/schematics/src/entity/schema.json:
```diff
@@ -49,6 +49,11 @@
         "Group actions, reducers and effects within relative subfolders",
       "aliases": ["g"]
+    },
+    "plural": {
```
This isn't needed, we always pluralize - https://github.com/ngrx/platform/blob/757d04ebeb35a8811c214a3b3c9e02318e650034/modules/schematics/src/entity/index.ts#L34

santoshyadav198613 replied:
removed

timdeschryver commented on modules/schematics/src/entity/schema.ts:
```diff
@@ -36,4 +36,9 @@ export interface Schema {
+   * Specifies if this is state and reducer variable will be plural
+   */
+
+  plural?: boolean;
```
Same here

santoshyadav198613 replied:
done

brandonroberts commented on modules/schematics-core/utility/ngrx-utils.ts:
```diff
@@ -90,11 +90,13 @@ export function addReducerToStateInterface(
     return new NoopChange();
   }
 
+  let state = stringUtils.camelize(options.name);
```
```suggestion
  let state = options.plural ? stringUtils.pluralize(options.name) : stringUtils.camelize(options.name);
```

Let's use the pluralize method instead of tacking on an `s`.

santoshyadav198613 replied:
done

brandonroberts commented on modules/schematics-core/utility/strings.ts:
```diff
@@ -110,6 +110,25 @@ export function capitalize(str: string): string {
+ 'innerHTML'.pluralize()         // 'InnerHTMLs'
+ 'action_name'.pluralize()       // 'actionNames'
+ 'css-class-name'.pluralize()    // 'cssClassNames'
+ 'regex'.capitalize()            // 'regexes'
```
```suggestion
 'regex'.pluralize()            // 'regexes'
```

santoshyadav198613 replied:
done

brandonroberts commented on modules/schematics-core/utility/strings.ts:
```diff
@@ -110,6 +110,25 @@ export function capitalize(str: string): string {
+ 'action_name'.pluralize()       // 'actionNames'
+ 'css-class-name'.pluralize()    // 'cssClassNames'
+ 'regex'.capitalize()            // 'regexes'
+ 'user'.capitalize()             // 'users'
```
```suggestion
 'user'.pluralize()             // 'users'
```

santoshyadav198613 replied:
done

brandonroberts commented on modules/schematics/schematics-core/utility/strings.ts:
```diff
@@ -110,6 +110,25 @@ export function capitalize(str: string): string {
 }
 
+/**
+ Returns the Pluralize form of a string
```
```suggestion
 Returns the plural form of a string
```

santoshyadav198613 replied:
done

santoshyadav198613 commented:
Hi @brandonroberts @timdeschryver ,
All changes are done please review.
